drivers/serial/uart_nrfx_uarte2.c

```diff
@@ -0,0 +1,1053 @@
+	const struct pinctrl_dev_config *pcfg;
+	uint32_t flags;
+
+	LOG_INSTANCE_PTR_DECLARE(log);
+};
+BUILD_ASSERT(offsetof(struct uarte_nrfx_config, a2i_config) == 0);
+
+#define UARTE_ERROR_FROM_MASK(mask)					\
+	((mask) & NRF_UARTE_ERROR_OVERRUN_MASK ? UART_ERROR_OVERRUN	\
+	 : (mask) & NRF_UARTE_ERROR_PARITY_MASK ? UART_ERROR_PARITY	\
+	 : (mask) & NRF_UARTE_ERROR_FRAMING_MASK ? UART_ERROR_FRAMING	\
+	 : (mask) & NRF_UARTE_ERROR_BREAK_MASK ? UART_BREAK		\
+	 : 0)
+
+static inline bool is_int_driven_api(const struct device *dev)
+{
+	const struct uarte_nrfx_config *config = dev->config;
+
+	return IS_ENABLED(CONFIG_UART_ASYNC_TO_INT_DRIVEN_API) &&
+		(config->flags & UARTE_CFG_FLAG_INTERRUPT_DRIVEN_API);
+}
+static inline bool is_sync_api(const struct device *dev)
+{
+	struct uarte_nrfx_data *data = dev->data;
+
+	return UARTE_ANY_POLLING && data->async == NULL;
+}
+
+static inline bool is_async_api(const struct device *dev)
+{
+	struct uarte_nrfx_data *data = dev->data;
+
+	return data->async && !is_int_driven_api(dev);
+}
+
+static inline const nrfx_uarte_t *get_nrfx_dev(const struct device *dev)
+{
+	const struct uarte_nrfx_config *config = dev->config;
+
+	return &config->nrfx_dev;
+}
+
+static int callback_set(const struct device *dev, uart_callback_t callback, void *user_data)
+{
+	struct uarte_nrfx_data *data = dev->data;
+
+	data->async->user_callback = callback;
+	data->async->user_data = user_data;
+
+	return 0;
+}
+
+#if UARTE_ANY_ASYNC
+static int api_callback_set(const struct device *dev, uart_callback_t callback, void *user_data)
+{
+	if (!is_async_api(dev)) {
+		return -ENOTSUP;
+	}
+
+	return callback_set(dev, callback, user_data);
+}
+#endif
+
+static void on_tx_done(const struct device *dev, const nrfx_uarte_event_t *event)
+{
+	struct uarte_nrfx_data *data = dev->data;
+	struct uart_event evt = {
+		.type = (event->data.tx.flags & NRFX_UARTE_TX_DONE_ABORTED) ?
+			UART_TX_ABORTED : UART_TX_DONE,
+		.data.tx.buf = event->data.tx.p_buffer,
+		.data.tx.len = event->data.tx.length
+	};
+	bool hwfc;
+
+#if CONFIG_UART_USE_RUNTIME_CONFIGURE
+	hwfc = data->uart_config.flow_ctrl == UART_CFG_FLOW_CTRL_RTS_CTS;
+#else
+	const struct uarte_nrfx_config *config = dev->config;
+
+	hwfc = config->nrfx_config.config.hwfc == NRF_UARTE_HWFC_ENABLED;
+#endif
+
+	if (hwfc) {
+		k_timer_stop(&data->async->tx_timer);
+	}
+	data->async->user_callback(dev, &evt, data->async->user_data);
+}
+
+static void on_rx_done(const struct device *dev, const nrfx_uarte_event_t *event)
+{
+	struct uarte_nrfx_data *data = dev->data;
+	const struct uarte_nrfx_config *config = dev->config;
+	struct uart_event evt;
+
+	if (event->data.rx.length) {
+		if (data->async->err) {
+			evt.type = UART_RX_STOPPED;
+			evt.data.rx_stop.reason = UARTE_ERROR_FROM_MASK(data->async->err);
+			evt.data.rx_stop.data.buf = event->data.rx.p_buffer;
+			evt.data.rx_stop.data.len = event->data.rx.length;
+			if (!(config->flags & UARTE_CFG_FLAG_INTERRUPT_DRIVEN_API)) {
```

Comment:
Actually I had to change those static inline to macros. That is because not all compilers are smart enough to determine from static inline that there is a dead code. There were linker issues when compiling for nrf52_bsim.